File: reto_clean_code/python/step4_informe_refactor.py
"""
Refactorización del ejercicio de Informe aplicando los principios SOLID.
Las responsabilidades se separan en interfaces y clases específicas para
generar informes, notificar y persistir la información.  Además, se
coordina el flujo desde un servicio que inyecta estas dependencias.
"""

from dataclasses import dataclass


class GeneradorInforme:
    def generar(self, data: 'InformeData') -> None:
        raise NotImplementedError()


class InformePDF(GeneradorInforme):
    def generar(self, data: 'InformeData') -> None:
        # Implementar generación de PDF
        pass


class Notificador:
    def notificar(self, data: 'InformeData') -> None:
        raise NotImplementedError()


class EmailNotificador(Notificador):
    def notificar(self, data: 'InformeData') -> None:
        # Implementar envío de correo electrónico
        pass


class Repositorio:
    def guardar(self, data: 'InformeData') -> None:
        raise NotImplementedError()


class InformeRepositorio(Repositorio):
    def guardar(self, data: 'InformeData') -> None:
        # Implementar persistencia en la base de datos
        pass


class InformeService:
    def __init__(self, generador: GeneradorInforme, notificador: Notificador, repositorio: Repositorio) -> None:
        self.generador = generador
        self.notificador = notificador
        self.repositorio = repositorio

    def procesar(self, data: 'InformeData') -> None:
        self.generador.generar(data)
        self.notificador.notificar(data)
        self.repositorio.guardar(data)


@dataclass
class InformeData:
    # Definir campos del informe según sea necesario
    contenido: str = ""

File: reto_clean_code/angular/step4-informe-refactor.ts
// Refactorización de la clase Informe aplicando principios SOLID.  Se
// definen interfaces y clases específicas para cada responsabilidad, y se
// utilizan en un servicio que orquesta el flujo.

// Interfaces
export interface GeneradorInforme {
  generar(data: any): void;
}

export interface Notificador {
  notificar(data: any): void;
}

export interface Repositorio {
  guardar(data: any): void;
}

// Implementaciones concretas
export class InformePdf implements GeneradorInforme {
  generar(data: any): void {
    // Implementar generación de PDF
  }
}

export class EmailNotificador implements Notificador {
  notificar(data: any): void {
    // Implementar notificación por correo
  }
}

export class InformeRepositorio implements Repositorio {
  guardar(data: any): void {
    // Implementar persistencia del informe
  }
}

// Servicio que coordina el proceso
export class InformeService {
  constructor(
    private generador: GeneradorInforme,
    private notificador: Notificador,
    private repositorio: Repositorio
  ) {}

  procesar(data: any): void {
    this.generador.generar(data);
    this.notificador.notificar(data);
    this.repositorio.guardar(data);
  }
}